[tokens] Fixes tokens being as powerful as passwords

dscape opened the pull request:
- Now tokens dont allow access to user document
- They allow very limited access to tokens itself

Needs review, we need to extend this everywhere to make sure tokens can only deploy apps.

dscape committed:
[tokens] better way to update, including feedback

dscape committed:
[tokens] Make tokens tokens, not passwords

dscape committed:
[tokens] removes access to organizations by token auth

dscape committed:
[tokens] narrow down permissions on token auth

* token auth is now useless for everything defined in http-users

dscape committed:
[tokens] adds test for /auth with token

dscape committed:
[tokens] whitespace pavan style

dscape commented:
@mmalecki this now prevents tokens from being used to access anything but `/auth`. However I have to check `nodejitsu/nodejitsu` to see if there's more stuff to protect (e.g. billing) +@chjj

dscape committed:
[tokens] adds tokens object to user

* allow to store third party tokens that can be used to call those
  services

chjj commented:
Yeah, I can limit some stuff in billing. If we're going to use this for jitsu, we probably don't need subscription creation or updating (yet).

Just an idea though. Since we auth on every request, do you think it might be better to define a hidden property with `Object.defineProperty` for `authMethod`? That way we don't have to mess with the User schema, and there's no chance it will actually be saved to the db. I might be wrong, but I don't see a reason to be saving this to the db.

dscape commented:
```
Just an idea though. 
Since we auth on every request, do you think it might be better to define a hidden property with Object.defineProperty for authMethod? 
That way we don't have to mess with the User schema, and there's no chance it will actually be saved to the db. 
I might be wrong, but I don't see a reason to be saving this to the db.
```

@chjj you are right but im too much of a JS n00b to know how to do this. Can you show and tell please? 

I tried without changing the schema and ended up with this not working (hence I added it). Storing in the database could be problematic, but only if a bug exists (previous auth was username and password and now something went wrong, so user got escalated privileges with a token)

chjj commented:
> Can you show and tell please?

Instead of:

``` js
user.authMethod = { method: "username/password" };
```

Maybe do:

``` js
      Object.defineProperty(user, 'authMethod', {
        value: { method: "username/password" },
        enumerable: false,
        configurable: true
      });
```

and then we wouldn't have to touch the schema, and the value wouldn't be stored in the db.

dscape committed:
[tokens] Allow all get requests (filtered)

* changes `requireUsernamePassword` mw to `requireUserPassForWrite`
  meaning any method expect GET
* orgs now require username password auth only for non gets
* users now require username password auth only for non gets
* user sensitive fields get filtered by default
* get /username/:who/tokens filters tokens that are not
  the currently authenticated one
* adds macros for valid authenticated user and valid token
  are reused because we have admins/non admins
* updates tests expectations accordingly

dscape committed:
[fix] removes `authMethod` from schema

* as per @chjj in https://github.com/flatiron/http-users/pull/22#issuecomment-10014736

dscape commented:
@chjj applied fix according to your code. tests pass :)

dscape committed:
[fix] s/user.tokens/user.vendorTokens/

dscape commented:
@3rd-Eden is this a good compromise for `webops` ?
